### PR TITLE
msg/async/dpdk: Fix the overflow while parsing dpdk coremask

### PR DESCRIPTION
**Pull request by rosinL**
coremask supports up to 256 bits in DPDK19.05, using stoull in
dpdk::eal:init limits coremask up to 64 bits. Parse coremask by hex
character from low to high.

Fixes: https://tracker.ceph.com/issues/43044

Signed-off-by: Hu Ye <yehu5@huawei.com>
Signed-off-by: Chunsong Feng <fengchunsong@huawei.com>
Signed-off-by: luo rixin <luorixin@huawei.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 1]
```diff
@@ -45,16 +45,47 @@ namespace dpdk {
     return std::bitset<CHAR_BIT * sizeof(n)>{n}.count();
   }
 
+  static int xdigit2val(unsigned char c)
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc]: s/xdigit2val/hex2dec/

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 2]
```diff
@@ -45,16 +45,47 @@ namespace dpdk {
...
+  {
+    int val;
+
+    if (isdigit(c))
+      val = c - '0';
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc]: nit, could just
```c++
return c - '0';
```

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 3]
```diff
@@ -45,16 +45,47 @@ namespace dpdk {
...
+    int count = 0;
+
+    if (coremask[0] == '0' && ((coremask[1] == 'x')
+        || (coremask[1] == 'X')))
+      coremask += 2;
+
+    for (int i = 0; coremask[i] != '\0'; i++) {
+      char c = coremask[i];
+      if (isxdigit(c) == 0)
+        return 0;
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc]: i think we need to return `-EINVAL` in this case.

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 4]
```diff
@@ -45,16 +45,47 @@ namespace dpdk {
...
+      int val = xdigit2val(c);
+      if (val)
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc]: nit, no need to make it a special case.

Comment (uniqueyehu) [reply src/msg/async/dpdk/dpdk_rte.cc]: If val is 0, it does not contain 1, so we will not count it.

Comment (tchaikov) [reply src/msg/async/dpdk/dpdk_rte.cc]: i think i understand what `bitcount()` is for. but my point is "no need to make it a special case", i.e. it does not make difference if we remove this `if`.

Comment (uniqueyehu) [reply src/msg/async/dpdk/dpdk_rte.cc]: get it, thanks

File: src/msg/async/dpdk/dpdk_rte.cc (hunk 3 above)
Comment (uniqueyehu) [reply src/msg/async/dpdk/dpdk_rte.cc]: The value it returns will be used to campare with `c->_conf->ms_async_op_threads` which is `unsigned`, considering that the returned value is only used in assert, so we choose to return 0 there. Is there any better suggestion?

Comment (tchaikov) [reply src/msg/async/dpdk/dpdk_rte.cc]: @uniqueyehu we can either continue using `ceph_assert()` to fail the daemon when running into invalid settings. or put something like

```c++
if (int bitcount = coremask_bitcount(coremask.c_str()); bitcount <= 0) {
  return -EINVAL;
} else if (static_cast<uint64_t>(bitcount) > c->_conf->ms_async_op_threads) {
  return -EINVAL;
}
```

File: src/msg/async/dpdk/dpdk_rte.cc [hunk 5]
```diff
@@ -45,18 +45,49 @@ namespace dpdk {
...
     bool done = false;
-    auto num = std::stoull(c->_conf.get_val<std::string>("ms_dpdk_coremask"),
-                           nullptr, 16);
-    unsigned int coremaskbit = bitcount(num);
+    auto coremask = c->_conf.get_val<std::string>("ms_dpdk_coremask");
+    int coremaskbit = coremask_bitcount(coremask.c_str());
 
-    ceph_assert(coremaskbit > c->_conf->ms_async_op_threads);
+    if (coremaskbit <= 0
+        || static_cast<uint64_t>(coremaskbit) <= c->_conf->ms_async_op_threads)
```
Comment (tchaikov) [inline src/msg/async/dpdk/dpdk_rte.cc]:
```suggestion
        || static_cast<uint64_t>(coremaskbit) < c->_conf->ms_async_op_threads)
```

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-kefu-testing-2019-12-20-1417/f36e245ae4f19a52641afb039f4ceeff043e3f26/

**Pull request merged**